soroban-sdk/src/testutils.rs

```diff
@@ -413,15 +415,60 @@ pub mod budget {
     }
 }
 
+#[derive(Clone)]
+pub struct ContractEvent {
+    pub env: Env,
+    pub contract_id: crate::Address,
+    pub topics: Vec<Val>,
+    pub data: Val,
+}
+
+impl Eq for ContractEvent {}
+
+impl PartialEq for ContractEvent {
+    fn eq(&self, other: &Self) -> bool {
+        if self.contract_id != other.contract_id || self.topics != other.topics {
+            false
+        } else if self.data.is_object() || other.data.is_object() {
+            self.env.obj_cmp(self.data, other.data).unwrap_infallible() == 0
+        } else {
+            self.data.get_payload() == other.data.get_payload()
+        }
+    }
+}
+
+impl Debug for ContractEvent {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(
+            f,
+            "ContractEvent {{ contract_id: {:?}, topics: {:?}, data: {:?} }}",
+            self.contract_id, self.topics, self.data
+        )
+    }
+}
+
+impl ContractEvent {
+    pub fn new(
+        env: &Env,
+        contract_id: crate::Address,
+        topics: Vec<Val>,
+        data: Val,
+    ) -> ContractEvent {
+        ContractEvent {
+            env: env.clone(),
+            contract_id,
+            topics,
+            data,
+        }
+    }
+}
+
 /// Test utilities for [`Events`][crate::events::Events].
 pub trait Events {
     /// Returns all events that have been published by contracts.
     ///
-    /// Returns a [`Vec`] of three element tuples containing:
-    /// - Contract ID
-    /// - Event Topics as a [`Vec<Val>`]
-    /// - Event Data as a [`Val`]
-    fn all(&self) -> Vec<(crate::Address, Vec<Val>, Val)>;
+    /// Returns a [`std::vec::Vec`] of [`ContractEvent`]s
+    fn all(&self) -> std::vec::Vec<ContractEvent>;
```

Comment:
Because this is a breaking change, I'm thinking: how could we do this so that it's not a breaking change? And/or how could we do this so that at least when people are using it to assert against some tuples, they don't have to change anything or aren't met with a whole bunch of errors and then have to figure out how to rewrite it all away from the tuples?

I think we should at very least add comparisons with tuples. That would hopefully make these OpenZeppelin example test cases pass and most existing uses pass.

Reply:
This was the thought behind #1634. It gets most of the way there in terms of testing semantics, but IMO does it in a less obvious way to prevent old tests from breaking.